src/Microsoft.Health.Fhir.Shared.Api/Controllers/ReindexController.cs

```diff
@@ -0,0 +1,207 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Threading.Tasks;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Serialization;
+using MediatR;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Api.Features.ActionResults;
+using Microsoft.Health.Fhir.Api.Features.Audit;
+using Microsoft.Health.Fhir.Api.Features.Filters;
+using Microsoft.Health.Fhir.Api.Features.Headers;
+using Microsoft.Health.Fhir.Api.Features.Routing;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Features.Context;
+using Microsoft.Health.Fhir.Core.Features.Operations;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.Api.Controllers
+{
+    [ServiceFilter(typeof(AuditLoggingFilterAttribute))]
+    [ServiceFilter(typeof(OperationOutcomeExceptionFilterAttribute))]
+    public class ReindexController : Controller
+    {
+        private readonly IMediator _mediator;
+        private readonly IFhirRequestContextAccessor _fhirRequestContextAccessor;
+        private readonly ReindexJobConfiguration _config;
+        private readonly ILogger<ReindexController> _logger;
+        private static Dictionary<string, HashSet<string>> _supportedParams = InitSupportedParams();
+
+        public ReindexController(
+            IMediator mediator,
+            IFhirRequestContextAccessor fhirRequestContextAccessor,
+            IOptions<OperationsConfiguration> operationsConfig,
+            ILogger<ReindexController> logger)
+        {
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(fhirRequestContextAccessor, nameof(fhirRequestContextAccessor));
+            EnsureArg.IsNotNull(operationsConfig?.Value?.Reindex, nameof(operationsConfig));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _mediator = mediator;
+            _fhirRequestContextAccessor = fhirRequestContextAccessor;
+            _config = operationsConfig.Value.Reindex;
+            _logger = logger;
+        }
+
+        [HttpPost]
+        [Route(KnownRoutes.Reindex)]
+        [ServiceFilter(typeof(ValidateReindexRequestFilterAttribute))]
+        [AuditEventType(AuditEventSubType.Reindex)]
+        public async Task<IActionResult> CreateReindexJob([FromBody] string body)
```

Comment:
Use do we use string instead of Resource here?

Reply:
Since the body is optional, it was a tricky to disambiguate between an empty body and a Resource parameter that failed to parse due to an invalid input.  Accepting a string and parsing it in the controller gave more control over validating the input.  But I do see that it would be nice to strongly type the input to the controller.

Reply:
Can you check ModelState to see if the input was invalid?